### PR TITLE
Tabulator session

### PR DESCRIPTION
**Pull request by moldover**
This PR adds a TabulatorSession object (issue #198) which manages the broader tabulation process including reading config, validating config, parsing cvr files, tabulating, and writing results.  This cleans up Main, and facilitates further testing and validation work.  For example, precinct IDs are now collected and stored in the TabulatorSession object.  This will allow #195 to be completed. 

I also added a bunch of regression tests based on existing test data.  With these tests we now have a real tool for protecting ourselves from making inadvertent changes. 

@tarheel @HEdingfield when this PR is merged, going forward, it will be the responsibility of anyone merging to master to ensure that all the tests complete successfully.  The current test suite runs in about 80 seconds on my computer.

### CODE REVIEW

File: src/main/java/com/rcv/TabulatorSession.java [hunk 1]
```diff
@@ -0,0 +1,260 @@
...
+      } catch (IOException exception) {
+        Logger.log(
+            Level.SEVERE, "Failed to configure tabulation logger!\n%s", exception.toString());
+      }
+
+      if (isTabulationLogSetUp) {
+        Logger.log(Level.INFO, "Logging tabulation to: %s", tabulationLogPath);
+        // Read cast vote records from CVR files
+        // parse the cast vote records
+        Pair<List<CastVoteRecord>, Set<String>> cvrInfo = parseCastVoteRecords(config);
```
Comment (tarheel) [inline src/main/java/com/rcv/TabulatorSession.java]: Not a big deal, but it's unfortunate that we then have to call `getValue` and `getKey` on this when it's not actually a key-value pair...

File: test_data/2015_portland_mayor_multi/2015_portland_mayor_multi_config.json [hunk 2]
```diff
@@ -0,0 +1,83 @@
+{
```
Comment (tarheel) [inline test_data/2015_portland_mayor_multi/2015_portland_mayor_multi_config.json]: I removed this one. It's not a real thing (mayor is not multi-seat) and it doesn't test anything in particular.

File: tests/com/rcv/TabulatorTests.java [hunk 3]
```diff
@@ -1,6 +1,6 @@
 /*
  * Ranked Choice Voting Universal Tabulator
- * Copyright (c) 2018 Jonathan Moldover, Louis Eisenberg, and Hylton Edingfield
+ * Copyright (c) 2018 Jonath  an Moldover, Louis Eisenberg, and Hylton Edingfield
```
Comment (tarheel) [inline tests/com/rcv/TabulatorTests.java]: undo

File: test_data/2015_portland_mayor_multi/2015_portland_mayor_multi_config.json (hunk 2 above)
Comment (moldover) [reply test_data/2015_portland_mayor_multi/2015_portland_mayor_multi_config.json]: Oh that's true.  I added this because I wanted a short and simple multi seat test for debugging something, but yeah it doesn't need to be a test.

File: src/main/java/com/rcv/TabulatorSession.java (hunk 1 above)
Comment (moldover) [reply src/main/java/com/rcv/TabulatorSession.java]: I changed it to pass the IDs as a parameter so it's a bit simpler.

File: src/main/java/com/rcv/TabulatorSession.java [hunk 4]
```diff
@@ -0,0 +1,260 @@
...
+    // set config file parent folder as default user folder
+    FileUtils.setUserDirectory(new File(configPath).getParent());
+    // config: the new contestConfig object
+    ContestConfig config = null;
+
+    // rawConfig holds the basic contest config data parsed from json
+    // this will be null if there is a problem loading it
+    RawContestConfig rawConfig = JsonParser.createRawContestConfigFromFile(configPath);
+
+    // if raw config failed alert user
```
[the remark this answers is not in the document]
Comment (moldover) [reply src/main/java/com/rcv/TabulatorSession.java]: I agree this code should be cleaned up (probably moved into ContestConfig), but it's outside the scope of this PR.  I created a new issue for it #210.

**Comment by moldover** [comment]
@tarheel I think I addressed all your comments

File: src/main/java/com/rcv/StreamingCVRReader.java [hunk 5]
```diff
@@ -185,6 +189,10 @@ private void endCVR() {
             currentRankings);
     // add it to overall list
     cvrList.add(newRecord);
+    // add precinct ID if one was found
+    if(currentPrecinct != null) {
```
Comment (tarheel) [inline src/main/java/com/rcv/StreamingCVRReader.java]: reminder to run Code > Reformat Code

File: src/main/java/com/rcv/TabulatorSession.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (tarheel) [reply src/main/java/com/rcv/TabulatorSession.java]: But yeah, I see now that you just moved this entire method from elsewhere, so it's reasonable not to modify it in this PR.

File: tests/com/rcv/TabulatorTests.java [hunk 6]
```diff
@@ -89,33 +90,36 @@ static boolean fileCompare(String path1, String path2) {
     return result;
   }
 
+  // function: getConfigPath
+  // purpose: given stem returns path to config file in test asset folder
+  // returns: path to config file
+  static String getConfigPath(String stem) {
```
Comment (HEdingfield) [inline tests/com/rcv/TabulatorTests.java]: You could generalize the `getConfigPath` function to generate the `expectedPath` too (e.g. with a simple string parameter that inserts whichever word after the underscore and before .json). You should rename the function if you do this though.

File: tests/com/rcv/TabulatorTests.java [hunk 7]
```diff
@@ -89,33 +90,36 @@ static boolean fileCompare(String path1, String path2) {
...
+
+    // create a session object and run the tabulation
+    TabulatorSession session = new TabulatorSession(TabulatorTests.getConfigPath(stem));
+    session.tabulate();
     // actualSummaryOutputPath is the summary json we just tabulated
-    String actualSummaryOutputPath = Main.getSummaryOutputPath();
+    String actualSummaryOutputPath = session.summaryOutputPath;
     // compare actual to expected
-    assertTrue(fileCompare(actualSummaryOutputPath, expectedPath));
+    assertTrue(fileCompare(expectedPath, actualSummaryOutputPath));
```
Comment (HEdingfield) [inline tests/com/rcv/TabulatorTests.java]: Since we're actually creating files here, I'd strongly recommend adding logic here that cleans up after the test is complete, e.g. a `teardown` function that removes all created files and directories after each test is run.

It's also really hard to parse out the results of each individual test given all the log / console window spam; I'd think the testing framework should allow us to only show the test results in a single output / console window? I guess it's not really a problem as long as we're running things inside of IntelliJ, but maybe something to add to the wishlist.

I'm not sure about the whole `junit.jupiter` sub-library; why are we using that instead of the basic `junit` suite?

File: tests/com/rcv/TabulatorTests.java [hunk 8]
```diff
@@ -75,9 +75,10 @@ static boolean fileCompare(String path1, String path2) {
         // both files have content so compare it
         if (!line1.equals(line2)) {
           // report inequality and keep processing in case there are more inequalities
-          Logger.log(Level.SEVERE, "files are not equal:\n%s\n%s", line1, line2);
+          Logger.log(Level.SEVERE, "files are not equal (%d)\n%s\n%s", currentLine, line1, line2);
           result = false;
```
Comment (HEdingfield) [inline tests/com/rcv/TabulatorTests.java]: Add a break here to avoid log spam if lots of lines aren't equal?

File: src/main/java/com/rcv/GuiConfigController.java [hunk 9]
```diff
@@ -243,7 +245,7 @@ public void buttonTabulateClicked() {
     if (checkForSaveAndTabulate()) {
       if (GuiContext.getInstance().getConfig() != null) {
         buttonBar.setDisable(true);
-        TabulatorService service = new TabulatorService();
+        TabulatorService service = new TabulatorService(selectedFile.getAbsolutePath());
```
[the remark this answers is not in the document]
Comment (moldover) [reply src/main/java/com/rcv/GuiConfigController.java]: I believe this is correct encapsulation.  TabulatorService needs to know about TabulationSession and how to construct one.  It does not need to know about GUIContext.  Why add that dependency?

File: src/main/java/com/rcv/StreamingCVRReader.java (hunk 5 above)
Comment (moldover) [reply src/main/java/com/rcv/StreamingCVRReader.java]: :+1:

File: tests/com/rcv/TabulatorTests.java [hunk 10]
```diff
@@ -75,9 +75,10 @@ static boolean fileCompare(String path1, String path2) {
         // both files have content so compare it
         if (!line1.equals(line2)) {
           // report inequality and keep processing in case there are more inequalities
-          Logger.log(Level.SEVERE, "files are not equal:\n%s\n%s", line1, line2);
+          Logger.log(Level.SEVERE, "files are not equal (%d)\n%s\n%s", currentLine, line1, line2);
```
[the remark this answers is not in the document]
Comment (moldover) [reply tests/com/rcv/TabulatorTests.java]: :+1:

File: tests/com/rcv/TabulatorTests.java (hunk 8 above)
Comment (moldover) [reply tests/com/rcv/TabulatorTests.java]: Added a counter which caps output at 10 lines

File: tests/com/rcv/TabulatorTests.java (hunk 6 above)
Comment (moldover) [reply tests/com/rcv/TabulatorTests.java]: dig it

File: tests/com/rcv/TabulatorTests.java (hunk 7 above)
Comment (moldover) [reply tests/com/rcv/TabulatorTests.java]: Yeah, I was thinking about something like that too... the scale test does produce a lot of data.  I think automated cleanup would be fine and just skip cleanup when the test fails so we can look at it.  I will add that.  

As for output, yes there is a visualizer for this.  Did you follow the directions here?  https://github.com/BrightSpots/rcv/pull/194

It is using junit5, and you should see something like this when running tests from IntelliJ:

regressiontests

File: src/main/java/com/rcv/GuiConfigController.java (hunk 9 above)
[the remark this answers is not in the document]
Comment (HEdingfield) [reply src/main/java/com/rcv/GuiConfigController.java]: It just seemed weird to me that we create a `TabulatorSession`object in line 178, but then we create a new one when we actually execute the tabulation. Maybe pass in the existing `TabulatorSession` instead of the path when initializing `TabulatorService`? Or maybe it's not a big deal? Just seemed like code smell to me.

File: tests/com/rcv/TabulatorTests.java (hunk 7 above)
Comment (HEdingfield) [reply tests/com/rcv/TabulatorTests.java]: Clean up sounds great!

Yep, sorry I wasn't more clear: I do have that working in IntelliJ, but I was thinking about someone running the tests outside of IntelliJ, like from the CLI. Just my reluctance to rely on third party software to develop effectively. Anyway, this isn't necessary to worry about now since the three of us are all using IntelliJ, but I think it's worth filing an issue to make it more friendly to other devs who might not be in the future.

File: src/main/java/com/rcv/GuiConfigController.java (hunk 9 above)
[the remark this answers is not in the document]
Comment (moldover) [reply src/main/java/com/rcv/GuiConfigController.java]: Gotcha.  Yes, your instinct is correct.  The code needs another piece of cleanup to refactor how ContestConfig objects are created and validated, so they can be used aside from TabulatorSession.  When that's done, the TabulatorSession reference in line 178 will be replaced with something like:
ContestConfig config = ContestConfig.loadContestConfig(fileToLoad.getAbsolutePath);

File: tests/com/rcv/TabulatorTests.java (hunk 7 above)
Comment (moldover) [reply tests/com/rcv/TabulatorTests.java]: Gotcha.  It really makes sense to use a thrid party tool here.  I researched this when I began planning the test code and there are various mature options.  I picked Junit5 because it is modern, widely used, well supported, and offers a rich feature set, although I expect we will only use the basics.  It can be integrated with other IDEs, and run standalone as well.  IMO it is pretty developer friendly.  https://stackoverflow.com/questions/52373469/how-to-launch-junit-5-platform-from-the-command-line-without-maven-gradle/52373592#52373592
